README.md
# nice_carousel

Build simple yet beautiful carousels for your mobile apps

## Features

* Carousel slides
* Carousel indicators
* Custom carousel content

![Demo Image](assets/demo.gif)

## Getting started

Add the following to pubspec.yaml:

```yaml
dependencies:
  nice_carousel:
```

## Usage

First, import the following files

```dart
import 'package:nice_carousel/carousel_slide.dart';
import 'package:nice_carousel/carousel_slides.dart';
```

Then, create a list of `CarouselSlide`:

```dart

List<CarouselSlide> carouselSlides = [
  CarouselSlide(
    title: '80% off',
    description: 'Free\ndiscount'.toUpperCase(),
    imageAsset: 'assets/img/1.webp',
  ),
  CarouselSlide(
    title: '60% off',
    description: 'Buy\nonline now'.toUpperCase(),
    imageAsset: 'assets/img/2.webp',
  ),
  CarouselSlide(
    title: '50% off',
    description: 'Grab\ndeals now'.toUpperCase(),
    imageAsset: 'assets/img/3.webp',
  ),
  CarouselSlide(
    title: '20% off',
    description: 'Shop\nonline now'.toUpperCase(),
    imageAsset: 'assets/img/4.webp',
  ),
];
```

You'll come up with a list of nice carousel slides

Finally, pass the slides to an instance of `CarouselSlides`

```dart

CarouselSlides carouselSlides = CarouselSlides(
  indicatorPosition: SlideIndicatorPosition.right,
  slides: slides,
  height: 300,
);

```

## Documentation

### CarouselSlide properties

| Name  |  Type |  Default | Description |
|---|---|---|---|
|  title | `String`  | required if CarouselSlide default constructor is used |Set the title of your carousel slide  |
| description  | `String`  | required if CarouselSlide default constructor is used | Set the description of your carousel slide  |
| imageAsset  | `String`  | required if CarouselSlide default constructor is used | Set the image of your carousel slide |
| networkAsset  |`String`   | required if CarouselSlide.network constructor is used | Set the network image path of your carousel slide |
| body  | `Widget` | required if CarouselSlide.custom constructor is used  | Set a custom body for your carousel |

### CarouselSlides properties

| Name  |  Type |  Default | Description |
|---|---|---|---|
| slides | `List<CarouselSlide>`  | required|list of your carousel slides |
| height  | `double`  | required | Set the height of your carousel slides  |
| indicatorColor  | `Color`  | Colors.white| Set the carousel's indicators colors |
| textStyle  |`TextStyle`  | null  | Set the text style of your carousel|
| indicatorPosition | `SlideIndicatorPosition`  |`SlideIndicatorPosition.right` |  Set the position of the indicators |
| hideIndicator | `bool`  |false |  Set whether the indicators is visible or not|

#### SlideIndicatorPosition enum properties

* SlideIndicatorPosition.left
* SlideIndicatorPosition.right

Now enjoy!

**Refer to [example folder](example/README.md) and the source code for more information.**

### Roadmap

More features coming soon 🔥

### Contact

Ethiel ADIASSA - [Twitter](https://www.twitter.com/enthusiastDev) - ethiel97@gmail.com
- [Website](https://ethieladiassa.me)

CHANGELOG.md
## [0.1.0] - [2022-06-06]

* Initial build for Flutter nice_carousel library


lib/carousel_slide.dart
import 'package:flutter/material.dart';

class CarouselSlide extends StatefulWidget {
  ///This is a builder for a carousel slide
  ///
  ///

  ///title of the carousel slide
  ///[String]
  String? title;

  ///description of the carousel slide
  ///[String]
  String? description;

  ///provide a custom content to your carousel slide
  ///[String]
  ///
  Widget? body;

  ///assets from your local assets directory
  ///[String]
  String? imageAsset;

  ///asset fetched over a network
  ///[String]
  String? networkAsset;

  ///[TextStyle]
  TextStyle? textStyle;

  ///[double]
  double? _transformAngle;

  ///[int]
  double? _transformScale;

  ///
  /// use this constructor if your image asset is provided from your assets directory
  ///
  CarouselSlide({
    super.key,
    required this.title,
    required this.description,
    required this.imageAsset,
    this.textStyle,
  }) {
    assert(networkAsset == null && imageAsset != null);
  }

  ///
  /// use this constructor if your image asset is fetched over a network
  ///
  CarouselSlide.network({
    super.key,
    required this.title,
    required this.description,
    required this.networkAsset,
    this.textStyle,
  }) {
    assert(imageAsset == null && networkAsset != null);
  }

  ///
  /// use this constructor if you want to provide a custom content for your carousel
  ///
  CarouselSlide.custom({
    super.key,
    required this.body,
    required this.imageAsset,
    this.textStyle,
  });

  @override
  State<CarouselSlide> createState() => _CarouselSlideState();
}

class _CarouselSlideState extends State<CarouselSlide> {
  NetworkImage? networkImage;
  AssetImage? assetImage;

  @override
  void initState() {
    if (widget.networkAsset != null) {
      networkImage = NetworkImage(widget.networkAsset!);
    }

    if (widget.imageAsset != null) {
      assetImage = AssetImage(widget.imageAsset!);
    }

    Future.delayed(Duration.zero, () {
      if (widget.networkAsset != null) {
        precacheImage(networkImage!, context);
      } else if (widget.imageAsset != null) {
        precacheImage(assetImage!, context);
      }
    });
    super.initState();
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
  }

  Widget get content =>
      widget.body != null ? buildCustomCarousel() : buildCarousel();

  DecorationImage get decorationImage => widget.imageAsset != null
      ? DecorationImage(
          colorFilter: ColorFilter.mode(
            Colors.black.withOpacity(.65),
            BlendMode.overlay,
          ),
          image: assetImage!,
          fit: BoxFit.cover,
        )
      : DecorationImage(
          colorFilter: ColorFilter.mode(
            Colors.black.withOpacity(.65),
            BlendMode.overlay,
          ),
          image: networkImage!,
          fit: BoxFit.cover,
        );

  Widget buildCarousel() => Container(
        padding: const EdgeInsets.all(
          14,
        ),
        width: double.infinity,
        decoration: BoxDecoration(
          backgroundBlendMode: BlendMode.overlay,
          color: Colors.black.withOpacity(
            .85,
          ),
          image: decorationImage,
        ),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Container(
              padding: const EdgeInsets.all(
                10,
              ),
              decoration: const BoxDecoration(
                color: Colors.white70,
              ),
              child: Text(
                "${widget.title}",
                textAlign: TextAlign.start,
                style: widget.textStyle?.apply(
                  fontWeightDelta: 2,
                  color: Colors.black87,
                ),
                maxLines: 1,
                overflow: TextOverflow.ellipsis,
              ),
            ),
            const SizedBox(
              height: 24,
            ),
            Text(
              widget.description!,
              textAlign: TextAlign.start,
              style: widget.textStyle?.apply(
                color: Colors.white,
                fontSizeDelta: 12,
                fontWeightDelta: 30,
              ),
              maxLines: 2,
              overflow: TextOverflow.ellipsis,
            ),
          ],
        ),
      );

  Widget buildCustomCarousel() => DecoratedBox(
        decoration: BoxDecoration(
          image: decorationImage,
        ),
        child: content,
      );

  @override
  Widget build(BuildContext context) => Transform(
        transform: /*(widget.transformAngle == 0 || widget.transformAngle == null)
            ? Matrix4.identity()
            : */
            Matrix4.identity()
              ..setEntry(3, 2, .001)
              ..rotateY(widget._transformAngle ?? 0),
        child: content,
      );
}


lib/carousel_slides.dart
import 'package:flutter/material.dart';
import 'package:nice_carousel/carousel_slide.dart';
import 'package:nice_carousel/carousel_slide_indicator.dart';

enum SlideIndicatorPosition {
  // center,
  right,
  left,
}

enum SlideIndicatorShape {
  circle,
  line,
}

enum SlideTransitionStyle {
  normal,
  translate,
}

class CarouselSlides extends StatefulWidget {
  CarouselSlides({
    Key? key,
    required this.slides,
    required this.height,
    this.indicatorColor = Colors.white,
    this.hideIndicator = false,
    this.textStyle,
    this.indicatorPosition = SlideIndicatorPosition.right,
  })  : assert(slides.isNotEmpty),
        super(key: key);

  ///
  /// whether the carousels indicator should be visible or not
  final bool hideIndicator;

  ///
  /// this is the main color of the indicators
  /// [Color]
  final Color? indicatorColor;

  /// this is the height of the carousel
  /// [String]
  final double height;

  ///
  /// this is the indicator of the carousel
  /// [SlideIndicatorPosition]
  ///
  final SlideIndicatorPosition indicatorPosition;

  ///
  /// list of carousel slides
  /// [List]
  ///
  final List<CarouselSlide> slides;

  ///
  /// this is the text style of the carousel
  /// [TextStyle]
  final TextStyle? textStyle;

  TextStyle get appTextStyle =>
      textStyle ??
      const TextStyle(
        fontFamily: 'packages/nice_carousel/Poppins',
        fontSize: 15,
        fontWeight: FontWeight.bold,
        color: Colors.white,
      );

  @override
  State<CarouselSlides> createState() => _CarouselSlidesState();
}

class _CarouselSlidesState extends State<CarouselSlides>
    with SingleTickerProviderStateMixin {
  late PageController pageController;

  late AnimationController animationController;

  int currentPage = 0;
  double pageOffset = 0;

  @override
  void initState() {
    super.initState();

    animationController = AnimationController(
        duration: const Duration(milliseconds: 500), vsync: this);

    pageController = PageController(initialPage: 0)
      ..addListener(() {
        pageOffset = pageController.page!;
      });
  }

  @override
  void dispose() {
    super.dispose();
    animationController.dispose();
    pageController.dispose();
  }

  @override
  Widget build(BuildContext context) => SizedBox(
        width: double.infinity,
        height: widget.height,
        child: Stack(
          children: [
            PageView(
              key: const Key("value"),
              controller: pageController,
              children: widget.slides.map(
                (carouselSlide) {
                  print("SLIDE INDEX $currentPage");
                  /*if (currentPage == pageOffset.floor()) {
                    // print("FROM PAGE ${pageOffset}\n");
                    return carouselSlide
                      ..transformAngle = pageOffset - currentPage;
                  } else if (currentPage == pageOffset.floor() + 1) {
                    // print("TO PAGE ${pageOffset}\n");

                    return carouselSlide
                      ..transformAngle = pageOffset - currentPage
                      ..transformScale = .8;
                  }*/
                  return carouselSlide
                    // ..transformAngle = 0
                    ..textStyle = widget.appTextStyle;
                },
              ).toList(),
              onPageChanged: (index) {
                setState(() {
                  currentPage = index;
                });
              },
            ),
            Visibility(
              visible: !widget.hideIndicator,
              child: Positioned(
                bottom: 18,
                right: widget.indicatorPosition == SlideIndicatorPosition.right
                    ? 14
                    : null,
                left: widget.indicatorPosition == SlideIndicatorPosition.left
                    ? 14
                    : null,
                child: Row(
                  mainAxisSize: MainAxisSize.max,
                  children: List.generate(widget.slides.length, (_) => (_))
                      .map(
                        (e) => CarouselSlideIndicator(
                          currentPage: currentPage == e,
                          color: widget.indicatorColor,
                        ),
                      )
                      .toList(),
                ),
              ),
            ),
          ],
        ),
      );
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"custom_carousel_flutter", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


lib/carousel_slide_indicator.dart
import 'package:flutter/material.dart';

class CarouselSlideIndicator extends StatelessWidget {
  final bool currentPage;

  final Color? color;

  const CarouselSlideIndicator({
    Key? key,
    this.color = Colors.white,
    required this.currentPage,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) => AnimatedContainer(
        duration: const Duration(
          milliseconds: 200,
        ),
        clipBehavior: Clip.none,
        margin: const EdgeInsets.all(2),
        height: currentPage ? 24 : 8,
        width: currentPage ? 6 : 8,
        decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(
              100,
            ),
            color:
                currentPage ? color?.withOpacity(.9) : color?.withOpacity(.5)
            // shape: BoxShape.circle,
            ),
      );
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:nice_carousel/carousel_slide.dart';
import 'package:nice_carousel/carousel_slides.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      color: Colors.white,
      title: 'Custom Carousel',
      home: Scaffold(
        body: ListView(
          children: [
            const SizedBox(
              height: 200,
            ),
            CarouselSlides(
              indicatorPosition: SlideIndicatorPosition.right,
              slides: [
                CarouselSlide(
                  title: '80% off',
                  description: 'Free\ndiscount'.toUpperCase(),
                  imageAsset: 'assets/img/1.webp',
                ),
                CarouselSlide(
                  title: '60% off',
                  description: 'Buy\nonline now'.toUpperCase(),
                  imageAsset: 'assets/img/2.webp',
                ),
                CarouselSlide(
                  title: '50% off',
                  description: 'Grab\ndeals now'.toUpperCase(),
                  imageAsset: 'assets/img/3.webp',
                ),
                CarouselSlide(
                  title: '20% off',
                  description: 'Shop\nonline now'.toUpperCase(),
                  imageAsset: 'assets/img/4.webp',
                ),
              ],
              height: 300,
            ),
          ],
        ),
      ),
    );
  }
}
